include/range/v3/view/set_algorithm.hpp

```diff
@@ -0,0 +1,904 @@
+
+                        return state_t::FIRST;
+                    }
+                    
+                    return state_t::ONLY_SECOND;
+                }
+                
+            public:
+                using value_type = common_type_t<range_value_t<R1>, range_value_t<R2>>;
+                using reference_type = common_reference_t<range_reference_t<R1>, range_reference_t<R2>>;
+                using rvalue_reference_type = common_reference_t<range_rvalue_reference_t<R1>,
+                                                                 range_rvalue_reference_t<R2>>;
+                using single_pass = meta::strict_or<SinglePass<range_iterator_t<R1>>,
+                                                    SinglePass<range_iterator_t<R2>>>;
+
+                set_union_cursor() = default;
+                set_union_cursor(pred_ref_ pred, proj1_ref_ proj1, proj2_ref_ proj2,
+                                 range_iterator_t<R1> it1, range_sentinel_t<R1> end1,
+                                 range_iterator_t<R2> it2, range_sentinel_t<R2> end2)
+                  : pred_(std::move(pred)), proj1_(std::move(proj1)), proj2_(std::move(proj2)),
+                    it1_(std::move(it1)), end1_(std::move(end1)), it2_(std::move(it2)), end2_(std::move(end2)),
+                    state(which_set())
+                {}
+                reference_type get() const
+                noexcept(noexcept(*it1_) && noexcept(*it2_))
+                {
+                    if(state == state_t::SECOND || state == state_t::ONLY_SECOND)
+                        return *it2_;
+                    else
+                        return *it1_;
+                }
+                void next()
+                {
+                    switch(state)
+                    {
+                        case state_t::FIRST:
+                            ++it1_;
+                            state = which_set();
+                            break;
+                            
+                        case state_t::FIRST_INC_SECOND:
+                            ++it2_;
+                            ++it1_;
+                            state = which_set();
+                            break;
+                        
+                        case state_t::ONLY_FIRST:
+                            ++it1_;
+                            break;
+                            
+                        case state_t::SECOND:
+                            ++it2_;
+                            state = which_set();
+                            break;
+                            
+                        case state_t::ONLY_SECOND:
+                            ++it2_;
+                            break;
+                    }
+                }
+                bool equal(set_union_cursor const &that) const
+                {
+                    return (it1_ == that.it1_ && it2_ == that.it2_) ||
+                           (done() && that.done());
```

Comment:
Putting on my semanticist hat: This implementation does *not* support comparing cursors into distinct ranges, because the resulting iterators fail to meet the requirements of `EqualityComparable`. To satisfy `EqualityComparable`, `==` has to mean substitutable. As a rule of thumb, any state that contributes to the behavior of an `EqualityComparable` type must either be accounted for by its equality operator, or objects that differ in the value of that state must be excluded from the domain of equality. (For readers familiar with the recent changes to the `Sentinel` concept in the TS, this argument was the cause.)

For the cursors defined here, significant state includes predicates, projection functions, and the beginnings and endings of the underlying ranges. Since the equality operators ignore predicates, projections, and the ends of the underlying ranges, cursors that don't agree in those attributes cannot be in the domain of equality. In other words, these operator definitions only support comparison of iterators into distinct ranges when those distinct ranges have the same predicate, projection functions, and ends-of-underlying ranges. If that is the case, then `this->done() && that.done()` implies `this->it1_ == that.it1_ && this->it2_ == that.it2_`; the `done` check is redundant.